### ISSUE TITLE
share_open_graph

### ISSUE THREAD

**Issue by efishman15** [jeduan/cordova-plugin-facebook4#74]
Hi Jeduan,
I'm using your great plugin inside my ionic/cordova app.
Currently built for Android.

All Facebook calls are working fine except the "showDialog" method specifically with the "share_open_dialog" method.

There is no much documentation on how does the JSON object should look like in this case, but from "sniffing" around your .java source code and facebook's source code, my JSON object looks like this:

{"method":"share_open_graph","action":{"og:type":"whosmarter:promote","team":"https://bnc.lt/m/THbuSY3HZn"},"previewPropertyName":"team"}

Nothing happens on screen when calling the facebookConnectPlugin.showDialog (no facebook animated gif icon) - and no error is returned.

The only thing I can see in the console logs of Android is:

W/PluginManager(19461): THREAD WARNING: exec() call to FacebookConnectPlugin.showDialog blocked the main thread for 184ms. Plugin should use CordovaInterface.getThreadPool().

I'm sure I am not passing the correct parameters - could you please assist with the proper documentation for this call scenario?

Using the facebookConnectPlugin.showDialog call with "share" or "feed" works just fine - as documented.

Thanks,
Eddy

**Comment by DoruLa** [comment]
Hi efishman15, I also would like to use the open graph stories with the share dialog. I want to share some custom stories. My solution was to slightly modify the plugin, on android (for now). Here's what I changed in the ConnectPlugin.java file:

Add the required import.
```java
import com.facebook.share.model.ShareOpenGraphObject;
```

Add `throws JSONException` to the executeDialog method.
```java
private void executeDialog(JSONArray args, CallbackContext callbackContext) throws JSONException {
```
Change the `else if (method.equalsIgnoreCase("share_open_graph"))` block into the following

```java
else if (method.equalsIgnoreCase("share_open_graph")) {
            if (!ShareDialog.canShow(ShareOpenGraphContent.class)) {
                callbackContext.error("Cannot show dialog");
                return;
            }
            showDialogContext = callbackContext;
            PluginResult pr = new PluginResult(PluginResult.Status.NO_RESULT);
            pr.setKeepCallback(true);
            showDialogContext.sendPluginResult(pr);

            if (!params.containsKey("action")) {
                callbackContext.error("Missing required parameter 'action'");
            }

            if (!params.containsKey("object")) {
                callbackContext.error("Missing required parameter 'object'.");
            }

            ShareOpenGraphObject.Builder objectBuilder = new ShareOpenGraphObject.Builder();
            JSONObject jObject = new JSONObject(params.get("object"));

            Iterator<?> keys = jObject.keys();

            String objectType = "";

            while ( keys.hasNext() ) {
                String key = (String)keys.next();
                String value = jObject.getString(key);

                objectBuilder.putString(key, value);

                if (key.equals("og:type"))
                    objectType = value;
            }

            if (objectType.equals("")) {
                callbackContext.error("Missing required object parameter 'og:type'");
            }

            ShareOpenGraphObject object = objectBuilder.build();

            ShareOpenGraphAction action = new ShareOpenGraphAction.Builder()
                    .setActionType(params.get("action"))
                    .putObject(objectType, object)
                    .build();

            ShareOpenGraphContent.Builder content = new ShareOpenGraphContent.Builder()
                    .setPreviewPropertyName(objectType)
                    .setAction(action);

            shareDialog.show(content.build());

        }
```

Now from cordova you can call:

```javascript
var options = {
                method: 'share_open_graph',
                action: 'appnamespace:action',
                object: '{"og:type":"appnamespace:objectname","og:title":"some title","og:url":"https://en.wikipedia.org/wiki/Main_Page","og:description":"A description"}'
            };

            facebookConnectPlugin.showDialog(options,
                function() {
                    window.alert('Working');
                },
                function(error) {
                    window.alert(JSON.stringify(error));
                }
            );
```

**NOTE**: I am not an Adroid or Java developper for that matter, so I am sure that the code can be optimized and maybe there are a few bugs, but it works for me, and maybe it will give you a good starting point.

**NOTE:** The above code works only for string properties on the object. If you want int, then you need to change the code. 

I hope this helps you. It's been a nuisance for me also to not be able to use open graph stories.

**Comment by jeduan** [comment]
@dorula interesting. Would you send a pull request?

**Comment by DoruLa** [comment]
@jeduan Done. I also added action properties, and tested them with int (expires_in) and datetime (starts_in) and it works. I also updated the documentation. I hope it is OK.

What I found during my testing, is that the callbacks (success or failure) are called only on the first call to showDialog. From the second call onward they don't get called anymore. So I suppose that some variables/values need to be reset? Maybe this helps you with issue #54.
